### PR TITLE
Add and use CProverString.ofCharArray

### PR DESCRIPTION
**Pull request by romainbrenguier**
This adds a function for converting char arrays to strings and use it
for a String constructor.

### CODE REVIEW

File: src/main/java/java/lang/StringBuilder.java [hunk 1]
```diff
@@ -241,23 +241,25 @@ public StringBuilder append(CharSequence s, int start, int end) {
...
         // return this;
-        return CProver.nondetWithNullForNotModelled();
     }
 
     /**
      * @throws IndexOutOfBoundsException {@inheritDoc}
-     *
-     * @diffblue.noSupport
+     * @diffblue.fullSupport
+     * @diffblue.untested
```
Comment (peterschrammel) [inline src/main/java/java/lang/StringBuilder.java:254]: How is this going to be tested?

**Review by peterschrammel — APPROVED** [review]

File: src/main/java/java/lang/StringBuilder.java (hunk 1 above)
Comment (romainbrenguier) [reply src/main/java/java/lang/StringBuilder.java:254]: It's tested by this PR https://github.com/diffblue/cbmc/pull/2374 but maybe we should have more tests in the models library.

File: src/main/java/java/lang/String.java [hunk 2]
```diff
@@ -216,11 +237,11 @@ public String(char value[]) {
      *          characters outside the bounds of the {@code value} array
      *
      * @diffblue.limitedSupport
```
Comment (allredj) [inline src/main/java/java/lang/String.java:239]: Why `limitedSupport` now?

File: src/main/java/org/cprover/CProverString.java [hunk 3]
```diff
@@ -388,4 +388,26 @@ public static StringBuffer insert(
         return CProver.nondetWithoutNullForNotModelled();
     }
 
+    /**
+     * Converts an array of characters to a string. This uses a loop and
+     * therefore in test-generation the {@code count} will be limited by the
+     * unwind parameter.
+     * This assumes {@code value} is not null, its length is greater or equal to
```
Comment (allredj) [inline src/main/java/org/cprover/CProverString.java]: `This assumes` is a bit ambiguous. It looks like a precondition to the function. Maybe `This constrains {@code value} to be not null...`?

File: src/main/java/org/cprover/CProverString.java [hunk 4]
```diff
@@ -388,4 +388,26 @@ public static StringBuffer insert(
...
+     * {@code offset + count} and {@code offset} and {@code count} are positive.
+     *
+     * @param value  Array of characters which is the source to copy from
+     * @param offset Index in {@code value} of the first character to copy
+     * @param count  Number of characters to copy
```
Comment (allredj) [inline src/main/java/org/cprover/CProverString.java]: We don't usually start the parameter description with an upper-case char.

File: src/main/java/org/cprover/CProverString.java [hunk 5]
```diff
@@ -388,4 +388,26 @@ public static StringBuffer insert(
         return CProver.nondetWithoutNullForNotModelled();
     }
 
+    /**
+     * Converts an array of characters to a string. This uses a loop and
+     * therefore in test-generation the {@code count} will be limited by the
+     * unwind parameter.
+     * This assumes {@code value} is not null, its length is greater or equal to
+     * {@code offset + count} and {@code offset} and {@code count} are positive.
```
Comment (allredj) [inline src/main/java/org/cprover/CProverString.java]: `positive` -> `non-negative`.

**Review by allredj — APPROVED** [review]
Only some documentation changes requested.

File: src/main/java/java/lang/StringBuilder.java (hunk 1 above)
Comment (romainbrenguier) [reply src/main/java/java/lang/StringBuilder.java]: It's tested by this PR https://github.com/diffblue/cbmc/pull/2374

Comment (romainbrenguier) [reply src/main/java/java/lang/StringBuilder.java]: but maybe we should add test in the models library

File: src/main/java/java/lang/String.java (hunk 2 above)
Comment (romainbrenguier) [reply src/main/java/java/lang/String.java:239]: The reasons are detailed below: count limited by unwind, and value cannot be null.

File: src/main/java/org/cprover/CProverString.java (hunk 4 above)
Comment (peterschrammel) [reply src/main/java/org/cprover/CProverString.java]: Hm. True, lower case seems standard in javadoc.

File: src/main/java/org/cprover/CProverString.java [hunk 6]
```diff
@@ -388,4 +388,26 @@ public static StringBuffer insert(
...
+     *
+     * @param value  Array of characters which is the source to copy from
+     * @param offset Index in {@code value} of the first character to copy
+     * @param count  Number of characters to copy
+     * @return The created String
+     */
+    public static String ofCharArray(char value[], int offset, int count) {
+        org.cprover.CProver.assume(value != null);
```
Comment (peterschrammel) [inline src/main/java/org/cprover/CProverString.java]: import

**Review by peterschrammel — APPROVED** [review]

**Pull request merged**